src/video_core/rasterizer_cache/rasterizer_cache.h

```diff
@@ -2,190 +2,1402 @@
+
+    Common::Rectangle<u32> depth_rect{};
+    SurfaceId depth_id{};
+    u32 depth_level{};
+    if (using_depth_fb)
+        std::tie(depth_id, depth_rect) = GetSurfaceSubRect(depth_params, ScaleMatch::Exact, false);
+
+    Common::Rectangle<u32> fb_rect{};
+    if (color_id && depth_id) {
+        fb_rect = color_rect;
+        // Color and Depth surfaces must have the same dimensions and offsets
+        if (color_rect.bottom != depth_rect.bottom || color_rect.top != depth_rect.top ||
+            color_rect.left != depth_rect.left || color_rect.right != depth_rect.right) {
+            color_id = GetSurface(color_params, ScaleMatch::Exact, false);
+            depth_id = GetSurface(depth_params, ScaleMatch::Exact, false);
+            fb_rect = slot_surfaces[color_id].GetScaledRect();
+        }
+    } else if (color_id) {
+        fb_rect = color_rect;
+    } else if (depth_id) {
+        fb_rect = depth_rect;
+    }
+
+    const Surface* color_surface = color_id ? &slot_surfaces[color_id] : nullptr;
+    const Surface* depth_surface = depth_id ? &slot_surfaces[depth_id] : nullptr;
+
+    if (color_id) {
+        color_level = color_surface->LevelOf(color_params.addr);
+        ValidateSurface(color_id, boost::icl::first(color_vp_interval),
+                        boost::icl::length(color_vp_interval));
+    }
+    if (depth_id) {
+        depth_level = depth_surface->LevelOf(depth_params.addr);
+        ValidateSurface(depth_id, boost::icl::first(depth_vp_interval),
+                        boost::icl::length(depth_vp_interval));
+    }
+
+    render_targets = RenderTargets{
+        .color_id = color_id,
+        .depth_id = depth_id,
+    };
+
+    return Framebuffer{runtime,     color_surface, color_level, depth_surface,
+                       depth_level, regs,          fb_rect};
+}
+
+template <class T>
+void RasterizerCache<T>::InvalidateFramebuffer(const Framebuffer& framebuffer) {
+    const auto invalidate = [&](SurfaceId surface_id) {
+        if (!surface_id) {
+            return;
+        }
+        Surface& surface = slot_surfaces[surface_id];
+        const SurfaceInterval interval = framebuffer.Interval(surface.type);
+        const PAddr addr = boost::icl::first(interval);
+        const u32 size = boost::icl::length(interval);
+        InvalidateRegion(addr, size, surface_id);
     };
+    const bool has_color = framebuffer.HasAttachment(SurfaceType::Color);
+    const bool has_depth = framebuffer.HasAttachment(SurfaceType::DepthStencil);
+    if (has_color) {
+        invalidate(render_targets.color_id);
+    }
+    if (has_depth) {
+        invalidate(render_targets.depth_id);
+    }
+}
+
+template <class T>
+typename RasterizerCache<T>::SurfaceRect_Tuple RasterizerCache<T>::GetTexCopySurface(
+    const SurfaceParams& params) {
+    Common::Rectangle<u32> rect{};
+
+    SurfaceId match_id = FindMatch<MatchFlags::TexCopy>(params, ScaleMatch::Ignore);
+
+    if (match_id) {
```

Comment:
consider returning std::optional 
```suggestion
    if (!match_id) {
        return std::nullopt;
    }
```

Reply:
The only place this function is used is in AccelerateTextureCopy which [checks](https://github.com/GPUCode/citra/blob/f40f49cfeb548215d6c72e803cbc2073736b3a4d/src/video_core/rasterizer_cache/rasterizer_cache.h#L137) the result immediately after so std::optional would be superflous here imo